File: same-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} p
# @param {TreeNode} q
# @return {Boolean}
def is_same_tree(p, q)
	ret = true
    if dfs(p, q) == false then 
    	ret = false
    end
    return ret
end

def dfs(root_p, root_q)
	if root_p == nil and root_q != nil then 
		return false
	elsif root_p != nil and root_q == nil then 
		return false
	elsif root_p == nil and root_q == nil
		return true
	else
		if root_p.val != root_q.val then
			return false
		else 
			return dfs(root_p.left, root_q.left) && dfs(root_p.right, root_q.right)
		end
	end
end

File: same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """

        def check(p, q):
            if not p and not q:
                return True
            if (p == None) ^ (q == None):
                return False
            return p.val == q.val and check(p.left, q.left) and check(p.right, q.right)

        return check(p, q)
